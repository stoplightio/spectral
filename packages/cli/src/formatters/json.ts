import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
