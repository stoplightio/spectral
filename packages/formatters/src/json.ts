import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    let documentationUrlObject = {};
    if (result.documentationUrl) {
      documentationUrlObject = {
        documentationUrl: result.documentationUrl,
      };
    }
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
      ...documentationUrlObject,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
