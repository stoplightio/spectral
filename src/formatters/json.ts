import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
