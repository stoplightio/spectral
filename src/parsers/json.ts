import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
