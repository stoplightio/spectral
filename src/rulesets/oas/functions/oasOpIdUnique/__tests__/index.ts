import { RuleType, Spectral } from '../../../../../index';
import { commonOasFunctions } from '../../../index';

import { rules } from '../../../index.json';

const ruleset = { functions: commonOasFunctions(), rules };

describe('oasOpIdUnique', () => {
  const s = new Spectral();

  s.addFunctions(ruleset.functions || {});
  s.addRules({
    'operation-operationId-unique': Object.assign(ruleset.rules['operation-operationId-unique'], {
      recommended: true,
      type: RuleType[ruleset.rules['operation-operationId-unique'].type],
    }),
  });

  test('validate a correct object', async () => {
    const results = await s.run({
      paths: {
        '/path1': {
          get: {
            operationId: 'id1',
          },
        },
        '/path2': {
          get: {
            operationId: 'id2',
          },
        },
      },
    });
    expect(results.length).toEqual(0);
  });

  test('return errors on different path operations same id', async () => {
    const results = await s.run({
      paths: {
        '/path1': {
          get: {
            operationId: 'id1',
          },
        },
        '/path2': {
          get: {
            operationId: 'id1',
          },
        },
      },
    });

    expect(results).toMatchSnapshot();
  });

  test('return errors on same path operations same id', async () => {
    const results = await s.run({
      paths: {
        '/path1': {
          get: {
            operationId: 'id1',
          },
          post: {
            operationId: 'id1',
          },
        },
      },
    });

    expect(results).toMatchSnapshot();
  });
});
