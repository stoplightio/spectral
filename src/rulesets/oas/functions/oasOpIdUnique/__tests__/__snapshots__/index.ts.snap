// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oasOpIdUnique return errors on different path operations same id 1`] = `
Array [
  Object {
    "code": "operation-operationId-unique",
    "message": "Every operation must have a unique \`operationId\`.",
    "path": Array [
      "paths",
      "/path1",
      "get",
      "operationId",
    ],
    "range": Object {
      "end": Object {
        "character": 28,
        "line": 4,
      },
      "start": Object {
        "character": 23,
        "line": 4,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "Every operation must have a unique \`operationId\`.",
  },
  Object {
    "code": "operation-operationId-unique",
    "message": "Every operation must have a unique \`operationId\`.",
    "path": Array [
      "paths",
      "/path2",
      "get",
      "operationId",
    ],
    "range": Object {
      "end": Object {
        "character": 28,
        "line": 9,
      },
      "start": Object {
        "character": 23,
        "line": 9,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "Every operation must have a unique \`operationId\`.",
  },
]
`;

exports[`oasOpIdUnique return errors on same path operations same id 1`] = `
Array [
  Object {
    "code": "operation-operationId-unique",
    "message": "Every operation must have a unique \`operationId\`.",
    "path": Array [
      "paths",
      "/path1",
      "get",
      "operationId",
    ],
    "range": Object {
      "end": Object {
        "character": 28,
        "line": 4,
      },
      "start": Object {
        "character": 23,
        "line": 4,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "Every operation must have a unique \`operationId\`.",
  },
  Object {
    "code": "operation-operationId-unique",
    "message": "Every operation must have a unique \`operationId\`.",
    "path": Array [
      "paths",
      "/path1",
      "post",
      "operationId",
    ],
    "range": Object {
      "end": Object {
        "character": 28,
        "line": 7,
      },
      "start": Object {
        "character": 23,
        "line": 7,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "Every operation must have a unique \`operationId\`.",
  },
]
`;
